// terminal-charts/src/core/Area.js
/**
 * Created by hustcc on 17/6/27.
 */

const invariant = require('../utils/invariant');
const types = require('../utils/types');
const Element = require('./Element');
const Layer = require('./Layer');

/**
 * 实心矩形
 * *****************************
 * *****************************
 * *****************************
 * *****************************
 * *****************************
 * *****************************
 */
class Area extends Element {
  constructor(start, end) {
    super();
    // 参数必须是 point 类型的
    invariant(
      types.isPoint(start) && types.isPoint(end),
      'TCharts: constructor props of Area should be (Point, Point), got (%s, %s).',
      types.typeof(start),
      types.typeof(end));
    // 校验通过，赋值属性
    this.start = start;
    this.end = end;

    this.box = {
      x1: Math.min(this.start.x, this.end.x),
      y1: Math.min(this.start.y, this.end.y),
      x2: Math.max(this.start.x, this.end.x),
      y2: Math.max(this.start.y, this.end.y),
    };
  }

  clone = () => new Area(this.start.clone(), this.end.clone());

  toString = () => `Area(${this.start}, ${this.end})`;

  draw = (options = { fill: '*' }) => {
    const fill = options.fill || '*';
    const width = this.box.x2 - this.box.x1;
    const height = this.box.y2 - this.box.y1;
    return new Layer(this.box, new Array(height).fill(new Array(width).fill(fill)));
  };
}

module.exports = Area;
