// terminal-charts/src/core/Line.js
/**
 * Created by hustcc on 17/6/27.
 */

const invariant = require('../utils/invariant');
const types = require('../utils/types');
const Element = require('./Element');
const Layer = require('./Layer');

/**
 * 一个线段
 * 横线
 * -------------------
 * 竖线
 * |
 * |
 * |
 * |
 * |
 * |
 */
class Line extends Element {
  constructor(start, end) {
    super();
    // 参数必须是 point 类型的
    invariant(
      types.isPoint(start) && types.isPoint(end),
      'TCharts: constructor props of Line should be (Point, Point), got (%s, %s).',
      types.typeof(start),
      types.typeof(end));
    // 控制台模式下，仅仅只能输入横线或者竖线
    invariant(
      start.x === end.x || start.y === end.y,
      'TCharts: we can only draw Horizontal / Vertical line in terminal. got(%s, %s).',
      start,
      end);
    // 校验通过，赋值属性
    this.start = start;
    this.end = end;
    this.box = {
      x1: Math.min(this.start.x, this.end.x),
      y1: Math.min(this.start.y, this.end.y),
      x2: Math.max(this.start.x, this.end.x),
      y2: Math.max(this.start.y, this.end.y),
    };
  }

  clone = () => new Line(this.start.clone(), this.end.clone());

  toString = () => `Line(${this.start}, ${this.end})`;

  draw = () => {
    const width = this.box.x2 - this.box.x1;
    const height = this.box.y2 - this.box.y1;

    let ascii = null;
    if (width !== 0) {
      // 横向线
      ascii = [new Array(width).fill('-')];
    } else {
      // 纵向线
      ascii = new Array(height).fill(['|']);
    }
    return new Layer(this.box, ascii, this.zIndex);
  };
}

module.exports = Line;
