// terminal-charts/src/core/Rect.js
/**
 * Created by hustcc on 17/6/27.
 */

const invariant = require('../utils/invariant');
const types = require('../utils/types');
const Element = require('./Element');
const Point = require('./Point');
const Line = require('./Line');
const Area = require('./Area');

/**
 *  带边框的填充矩形，在控制台下面，仅仅只能绘制四边形
 *
 *  +---------------------------+
 *  |***************************|
 *  |***************************|
 *  |***************************|
 *  |***************************|
 *  +---------------------------+
 *
 */
class Rect extends Element {
  constructor(start, end) {
    super();
    // 参数必须是 point 类型的
    invariant(
      types.isPoint(start) && types.isPoint(end),
      'TCharts: constructor props of Rect should be (Point, Point), got (%s, %s).',
      types.typeof(start),
      types.typeof(end));
    // 校验通过，赋值属性
    this.start = start;
    this.end = end;
    this.box = {
      x1: Math.min(this.start.x, this.end.x),
      y1: Math.min(this.start.y, this.end.y),
      x2: Math.max(this.start.x, this.end.x),
      y2: Math.max(this.start.y, this.end.y),
    };
  }

  clone = () => new Rect(this.start.clone(), this.end.clone());

  toString = () => `Rect(${this.start}, ${this.end})`;

  draw = (options = { fill: '*' }) => {
    const area = new Area(this.start, this.end).draw(options);

    // 四个顶点
    const points = [
      new Point(this.box.x1, this.box.y1),
      new Point(this.box.x2, this.box.y2),
      new Point(this.box.x1, this.box.y2),
      new Point(this.box.x2, this.box.y1),
    ];
    const corners = [];
    points.forEach((p) => {
      corners.push(p.draw(options));
    });

    // 四条边
    const lines = [
      new Line(points[0], points[1]).draw(options),
      new Line(points[1], points[2]).draw(options),
      new Line(points[2], points[3]).draw(options),
      new Line(points[3], points[0]).draw(options),
    ];
    // 合并三个部分
    return area.mergeArray(lines.concat(corners));
  };
}

module.exports = Rect;
