// terminal-charts/src/core/Text.js
/**
 * Created by hustcc on 17/6/27.
 */

const evenly = require('evenly');
const WordWidth = require('word-width');
const invariant = require('../utils/invariant');
const types = require('../utils/types');
const Element = require('./Element');
const Layer = require('./Layer');

/**
 * 一段文本（目前仅支持横向）
 *
 * hello world，这个是一个控制台图标库
 */
class Text extends Element {
  constructor(p, text) {
    super();
    invariant(
      types.isPoint(p) && types.isString(text),
      'TCharts: constructor props of Line should be (Point, string), got (%s, %s).',
      types.typeof(p),
      types.typeof(text));

    this.p = p;
    this.text = text;

    // 居中显示
    const _width = evenly(WordWidth(this.text), 2, 0);
    this.box = {
      x1: p.x - _width[0],
      y1: p.y,
      x2: p.x + _width[1],
      y2: p.y,
    };
  }

  clone = () => new Text(this.p, this.text);

  toString = () => `Text(${this.p}, ${this.text})`;

  draw = () => {
    const width = this.x2 - this.x1;
    const textLine = [this.text].concat(new Array(width - 1).fill(''));
    return new Layer(this.box, [textLine]);
  };
}

module.exports = Text;
